qualtran/bloqs/arithmetic/permutation.py

```diff
@@ -0,0 +1,248 @@
+
+            delta = x_k ^ self.cycle[(k + 1) % len(self.cycle)]
+            a, x = bb.add_t(XorK(QUInt(self.bitsize), delta).controlled(), ctrl=a, x=x)
+
+        x, a = bb.add_t(EqualsAConstant(self.bitsize, self.cycle[0]), x=x, target=a)
+
+        bb.free(cast(Soquet, a))
+
+        return {'x': x}
+
+    def build_call_graph(self, ssa: 'SympySymbolAllocator') -> Set['BloqCountT']:
+        if self.is_symbolic():
+            x = ssa.new_symbol('x')
+            cycle_len = slen(self.cycle)
+            return {
+                (EqualsAConstant(self.bitsize, x), cycle_len + 1),
+                (XorK(QUInt(self.bitsize), x).controlled(), cycle_len),
+            }
+
+        return super().build_call_graph(ssa)
+
+
+@bloq_example
+def _permutation_cycle() -> PermutationCycle:
+    permutation_cycle = PermutationCycle(4, (0, 1, 2))
+    return permutation_cycle
+
+
+@bloq_example
+def _permutation_cycle_symb() -> PermutationCycle:
+    import sympy
+
+    N, L = sympy.symbols("N L")
+    cycle = Shaped((L,))
+    permutation_cycle_symb = PermutationCycle(N, cycle)
+    return permutation_cycle_symb
+
+
+def _convert_cycles(cycles) -> Union[tuple[SymbolicCycleT, ...], Shaped]:
+    if isinstance(cycles, Shaped):
+        return cycles
+    return tuple(_convert_cycle(cycle) for cycle in cycles)
+
+
+@frozen
+class Permutation(Bloq):
+    """Apply a permutation of [0, N - 1] on the basis states.
+
+    Decomposes a permutation into cycles and applies them in order.
+    See :meth:`from_dense_permutation` to construct this bloq from a permutation.
+
+    Args:
+        N: the total size the permutation acts on.
+        cycles: a sequence of permutation cycles that form the permutation.
+
+    Registers:
+        x: integer register storing a value in [0, ..., N - 1]
+
+    References:
+        [A simple quantum algorithm to efficiently prepare sparse states](https://arxiv.org/abs/2310.19309v1)
+        Appendix B.
+    """
+
+    N: SymbolicInt
+    cycles: Union[tuple[SymbolicCycleT, ...], Shaped] = field(converter=_convert_cycles)
+
+    @cached_property
+    def signature(self) -> Signature:
+        return Signature.build_from_dtypes(x=BoundedQUInt(self.bitsize, self.N))
+
+    @cached_property
+    def bitsize(self):
+        return bit_length(self.N - 1)
+
+    def is_symbolic(self):
+        return is_symbolic(self.N, self.cycles) or (
+            isinstance(self.cycles, tuple) and is_symbolic(*self.cycles)
+        )
+
+    @classmethod
+    def from_dense_permutation(cls, permutation: Sequence[int]):
+        N = len(permutation)
+        cycles = tuple(decompose_permutation_into_cycles(permutation))
+        return cls(N, cycles)
+
+    @classmethod
+    def from_sparse_permutation_prefix(cls, N: int, permutation_prefix: Sequence[int]):
+        cycles = tuple(decompose_sparse_prefix_permutation_into_cycles(permutation_prefix, N))
+        return cls(N, cycles)
```

Comment:
Please add a docstring explaining `permutation_prefix` - copying the docstring from `decompose_sparse_prefix_permutation_into_cycles` should suffice. 

For specifying a sparse permutation, I would naturally think of specifying a dictionary instead of a sequence of prefix but this also LGTM % adding docstrings for clarity

Reply:
used a dictionary instead, actually made the code simpler!